services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/NetworkIPAvailabilityServiceImpl.java

```diff
@@ -0,0 +1,215 @@
+        distributed under the License is distributed on an "AS IS" BASIS,
+        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+package com.futurewei.alcor.vpcmanager.service.Impl;
+
+import com.futurewei.alcor.common.db.CacheException;
+import com.futurewei.alcor.common.exception.ResourceNotFoundException;
+import com.futurewei.alcor.common.exception.ResourcePersistenceException;
+import com.futurewei.alcor.vpcmanager.service.NetworkIPAvailabilityService;
+import com.futurewei.alcor.vpcmanager.service.VpcDatabaseService;
+import com.futurewei.alcor.vpcmanager.utils.IPUtil;
+import com.futurewei.alcor.web.entity.subnet.AllocationPool;
+import com.futurewei.alcor.web.entity.subnet.SubnetEntity;
+import com.futurewei.alcor.web.entity.subnet.SubnetWebJson;
+import com.futurewei.alcor.web.entity.vpc.NetworkIPAvailabilityEntity;
+import com.futurewei.alcor.web.entity.vpc.SubnetIPAvailabilityEntity;
+import com.futurewei.alcor.web.entity.vpc.VpcEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Service
+public class NetworkIPAvailabilityServiceImpl implements NetworkIPAvailabilityService {
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @Autowired
+    VpcDatabaseService vpcDatabaseService;
+
+    @Value("${microservices.subnet.service.url}")
+    private String subnetUrl;
+
+    private RestTemplate restTemplate = new RestTemplate();
+
+    @Override
+    public NetworkIPAvailabilityEntity getNetworkIPAvailability(String vpcid) throws ResourceNotFoundException, ResourcePersistenceException {
+
+        NetworkIPAvailabilityEntity networkIPAvailabilityEntity = new NetworkIPAvailabilityEntity();
+
+        VpcEntity vpcEntity = this.vpcDatabaseService.getByVpcId(vpcid);
+        if (vpcEntity == null) {
+            throw new ResourceNotFoundException("vpc can not be found");
+        }
+
+        networkIPAvailabilityEntity = setNetworkIPAvailability(vpcEntity);
+
+        return networkIPAvailabilityEntity;
+    }
+
+    @Override
+    public List<NetworkIPAvailabilityEntity> getNetworkIPAvailabilities(String vpcId, String vpcName, String tenantId, String projectId) throws CacheException, ResourceNotFoundException {
+
+        List<NetworkIPAvailabilityEntity> networkIPAvailabilityEntities = new ArrayList<NetworkIPAvailabilityEntity>();
+
+        Map<String, VpcEntity> vpcStates = this.vpcDatabaseService.getAllVpcs();
+
+        // Filter vpcStates by request parameters
```

Comment:
This part could become slow when the number of VPC exceed certain threshold. It loads all the vpc states to the memory from DB cache/storage, and filter them on app level, not the most efficient way.

Pls check the LIST method (like LIST vpc, subnet, port etc.) and use database query. @kevin-zhonghao

Reply:
Got it